### PR TITLE
API: Remove `seterrobj`/`geterrobj`/`extobj=` and related C-API functions

### LINKED ISSUE #23900: API: Slash down errstate related API to make it sustainable
For NumPy 2.0 (or well, TBH, I think we could also pull this off in a normal release, but then I might ask for a brief NEP, because I want to jump deprecations), I would like:

* Delete any weird "log" and "callback" stuff from the errobj/errstate
* Remove `seterrobj`, `geterrobj`, `seterrcall`, `geterrcall` *and* the ufunc `errobj=`
  * I would add an error on the last one with a note that we are willing to accept patches to restore some of its behavior if there is a real need.  I don't hate allowing passing this state in, but I think it should evolve (and probably have a new name).
* `geterr()` and `seterr()` are OK, since they take/return dicts.  `errstate` is preferable, but they don't matter for this.  (I am OK with removing them, but I think just deprecating them is also fine.)

`errobj` is a weird mutable list, I don't think using it is good and I suspect it will just make fixing the errstate hard.

Now the callback is probably used a couple of time in the wild, I found this: https://github.com/aglavic/genx/blob/8083fb73271f24dbf8c92433a6a86b422fefede5/genx/genx/core/custom_logging.py#L80

I think it should be fine (and backwards compatible) to ask such programs to monkey-patch [`warnings.showwarning`](https://docs.python.org/3/library/warnings.html#warnings.showwarning) instead.

---

Can I get a few :+1: for this change? :)

*EDIT: Whooops, the ufunc keyword argument is called `extobj` not `errobj`, shows how much its used ;).*

### PR DESCRIPTION
**Pull request by seberg**
This has a couple of different commits and is unfortunately a bit large because of that.  With the exception of the necessary linalg cleanup, they are rather intertwined though.

The commits should be cleanly reviewable one-by-one, though!

*Not marking as draft, but the release note build may error until gh-23919 is merged.*

Closes gh-23900.  There were 2 :+1: on the issue, so I assume/hope we have a consensus on these.  Right now this is the most minimal useful version:  Remove everything directly exposing `extobj` or the C-API side tuple of `(ufunc_name, callable)` which is a weird indirection.

~EDIT: OK, removing "log" and "print" is not minimal.~

### CODE REVIEW

File: numpy/core/src/umath/extobj.c [hunk 1]
```diff
@@ -18,6 +18,45 @@
...
+    if (errmask && retstatus) {
+        HANDLEIT(DIVIDEBYZERO, "divide by zero");
+        HANDLEIT(OVERFLOW, "overflow");
+        HANDLEIT(UNDERFLOW, "underflow");
+        HANDLEIT(INVALID, "invalid value");
+    }
+    return 0;
+}
+
+#undef HANDLEIT
```
Comment (seberg) [inline numpy/core/src/umath/extobj.c:57]: This is a straight copy paste, the only change is making them `static`.

**Comment by seberg** [comment]
@mattip I saw some of those unpacking non 0-d arrays into scalars in the f2py tests here also.  Maybe something unrelated to the meson PR changed?!

**Comment by seberg** [comment]
I have dropped removal of `'log'` and `'print'` mode.  It shouldn't matter much for introducing a `ContextVar` (I wouldn't mind removing at least `print`, but no need to do it here).

Technically, the C-API could be retained.  But I still think it will make the refactor unnecessarily annoying as it will require duplicate code for only that purpose and there is no known user.

**Comment by mattip** [comment]
Thanks @seberg. Nice cleanup.

**Pull request merged**